build-system/tasks/cherry-pick.js

```diff
@@ -0,0 +1,179 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const argv = require('minimist')(process.argv.slice(2));
+const log = require('fancy-log');
+const {getOutput} = require('../common/exec');
+const {green, cyan, red, yellow} = require('ansi-colors');
+
+const RTV_PREFIXES = {
+  '00': 'experimental',
+  '01': 'stable',
+  '02': 'control',
+  '03': 'beta',
+  '04': 'nightly',
+  '05': 'nightly-control',
+};
+
+/**
+ * Executes a shell command, and logs an error message if the command fails.
+ *
+ * @param {string} cmd
+ * @param {string} msg
+ * @return {!Object}
+ */
+function execOrLog(cmd, msg) {
+  const result = getOutput(cmd);
+  if (result.status) {
+    log(yellow('ERROR:'), msg);
+    throw new Error(result.stderr);
+  }
+
+  return result;
+}
+
+/**
+ * Determines the name of the cherry-pick branch.
+ *
+ * @param {string} rtv
+ * @param {Array<string>} commits
+ * @return {string}
+ */
+function cherryPickBranchName(rtv, commits) {
+  const prefix = rtv.substr(0, 2);
+  const channel = RTV_PREFIXES[prefix] || 'unknown';
+  const [date] = new Date().toISOString().split('T');
+
+  return `amp-release-${date}-${channel}+${commits.length}`;
+}
+
+/**
+ * Updates tags from the remote and creates a branch at the release commit.
+ *
+ * @param {string} ref
+ * @param {string} branch
+ * @param {string} remote
+ */
+function prepareBranch(ref, branch, remote) {
+  const {status} = getOutput(`git rev-parse ${ref}`);
+  // Skip fetching remote tags if the tag is already available.
+  if (status === 0) {
+    log(green('INFO:'), 'Identified ref', cyan(ref), 'in local repository');
+  } else {
+    log(green('INFO:'), 'Fetching latest tags from', cyan(remote));
+    execOrLog(
+      `git fetch ${remote} "refs/tags/*:refs/tags/*"`,
+      `Failed to fetch updated tags from remote ${cyan(remote)}`
+    );
+  }
+
+  execOrLog(
+    `git checkout -b ${branch} ${ref}`,
+    `Failed to checkout new branch at ref ${cyan(ref)}`
+  );
+}
+
+/**
+ * Cherry-picks a commit into a new branch. When the cherry-pick succeeds,
+ * returns `true`. In the event of a merge conflict, the cherry-pick is aborted
+ * and an error is thrown.
+ *
+ * @param {string} sha
+ */
+function performCherryPick(sha) {
+  try {
+    log(green('INFO:'), 'Cherry-picking commit', cyan(sha));
+    execOrLog(
+      `git cherry-pick -x ${sha}`,
```

Comment:
General point for consideration: All our `git` logic currently lives in `build-system/common/git.js` so it can be maintained in one place. If it makes sense, maybe move the building block functions in this file to `git.js`? (Unless you think there's too much code to move, and this file is all about `git` operations in any case.)

Reply:
Everything in `git.js` returns the output. In our case, we need the actual status code/returned process object. I see no overlap in functionality between the two files, and as you said this file is all git operations anyway, so I'm going to opt for the parethesized option here and leave it separate.